src/allmydata/web/unlinked.py

```diff
@@ -59,34 +73,70 @@ def _done(upload_results, redir_to):
     return d
 
 
-class UploadResultsPage(status.UploadResultsRendererMixin, rend.Page):
+class UploadResultsPage(MultiFormatResource):
     """'POST /uri', to create an unlinked file."""
-    docFactory = getxmlfile("upload-results.xhtml")
 
     def __init__(self, upload_results):
-        rend.Page.__init__(self)
-        self.results = upload_results
+        """
+        :param IUploadResults upload_results: stats provider.
+        """
+        super(UploadResultsPage, self).__init__()
+        self._upload_results = upload_results
+
+    def render_HTML(self, req):
+        elem = UploadResultsElement(self._upload_results)
+        return renderElement(req, elem)
+
+    # This is weird but necessary because:
+    #
+    #  1. MultiFormatResource.render() uses argument "t" to figure out
+    #     its output format.
+    #
+    #  2. Upload request is of the form "POST /uri?t=upload&file=newfile".
+    #     See URIHandler.render_POST().
+    #
+    # MultiFormatResource.render() looks up "t" argument, which in
+    # this case has the value "upload", and then it would look for a
+    # render_UPLOAD() method.
+    #
+    # We could change upload request to use more descriptive names
+    # that do not cause name collisions like this.  That should be a
+    # separate change though.
+    render_UPLOAD = render_HTML
+
+
+class UploadResultsElement(status.UploadResultsRendererMixin):
+
+    loader = XMLFile(FilePath(__file__).sibling("upload-results.xhtml"))
+
+    def __init__(self, upload_results):
+        super(UploadResultsElement, self).__init__()
+        self._upload_results = upload_results
 
     def upload_results(self):
-        return defer.succeed(self.results)
+        return defer.succeed(self._upload_results)
 
-    def data_done(self, ctx, data):
+    @renderer
+    def done(self, req, tag):
         d = self.upload_results()
         d.addCallback(lambda res: "done!")
         return d
 
-    def data_uri(self, ctx, data):
+    @renderer
+    def uri(self, req, tag):
         d = self.upload_results()
         d.addCallback(lambda res: res.get_uri())
         return d
 
-    def render_download_link(self, ctx, data):
+    @renderer
+    def download_link(self, req, tag):
         d = self.upload_results()
         d.addCallback(lambda res:
-                      T.a(href="/uri/" + urllib.quote(res.get_uri()))
-                      ["/uri/" + res.get_uri()])
+                      tags.a("/uri/" + res.get_uri(),
+                             href="/uri/" + urllib.quote(res.get_uri())))
```

Comment:
This should probably use the `hyperlink` library to build up the URL since we already depend on it (although, I guess also this is the same code as before so maybe we should just leave it alone for now...)

Reply:
What is the advantage in using `hyperlink`?  I guess we can leave this alone for now, unless using `hyperlink` is clearly the right thing to do.

Reply:
It's more likely to produce a valid URL than adding strings together ... but, yes, lets leave this alone in this porting-PR :)